File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-0.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # Use 128 to cover all standard ASCII characters (not just 'a-z')
        last_seen = [-1] * 128 
        left = 0
        max_len = 0
        
        for right in range(len(s)):
            char_code = ord(s[right])
            
            # If we've seen this char and it's within our current window
            if last_seen[char_code] >= left:
                # Jump 'left' to one position after the previous occurrence
                left = last_seen[char_code] + 1
            
            # Update the last seen index for this character
            last_seen[char_code] = right
            
            # Update the max length found so far
            # (right - left + 1) is the length of the current window
            max_len = max(max_len, right - left + 1)
            
        return max_len

File: Data Structures & Algorithms/burst-balloons/submission-0.cpp
class Solution {
public:
    int maxCoins(std::vector<int>& nums) {
        int n = nums.size();
        
        std::vector<int> arr(n + 2, 1);
        for (int i = 0; i < n; ++i) {
            arr[i + 1] = nums[i];
        }
        
        std::vector<std::vector<int>> dp(n + 2, std::vector<int>(n + 2, 0));
        
        for (int window = 1; window <= n; ++window) {
            for (int left = 1; left <= n - window + 1; ++left) {
                int right = left + window - 1;
                
                for (int k = left; k <= right; ++k) {
                    int coins = arr[left - 1] * arr[k] * arr[right + 1];
                    int total = coins + dp[left][k - 1] + dp[k + 1][right];
                    dp[left][right] = std::max(dp[left][right], total);
                }
            }
        }
        
        return dp[1][n];
    }
};
